File: typescript/src/algorithms/position-run-detection.ts
import { DraftPick, PositionRun } from '../types';
import { config } from '../config';

/**
 * Detect if a position run is occurring
 * A run is detected when 3+ consecutive picks are the same position
 */
export function detectPositionRun(
  recentPicks: DraftPick[],
  threshold: number = config.algorithms.positionRunThreshold
): PositionRun {
  if (recentPicks.length < threshold) {
    return { detected: false };
  }

  const lastN = recentPicks.slice(-threshold);
  const positions = lastN.map(p => getPlayerPosition(p.player_id)); // TODO: Implement player lookup

  // Find most common position in recent picks
  const positionCounts: Record<string, number> = {};
  positions.forEach(pos => {
    positionCounts[pos] = (positionCounts[pos] || 0) + 1;
  });

  const mostCommon = Object.keys(positionCounts).reduce((a, b) =>
    positionCounts[a] > positionCounts[b] ? a : b
  );

  if (positionCounts[mostCommon] >= threshold) {
    return {
      detected: true,
      position: mostCommon,
      count: positionCounts[mostCommon],
    };
  }

  return { detected: false };
}

/**
 * Adjust scarcity multiplier based on position run
 * +15% per pick beyond threshold
 */
export function adjustScarcity(position: string, runCount: number): number {
  const baseMultiplier = config.algorithms.scarcityMultipliers[position as keyof typeof config.algorithms.scarcityMultipliers] || 1.0;
  const threshold = config.algorithms.positionRunThreshold;

  if (runCount <= threshold) {
    return baseMultiplier;
  }

  const additionalMultiplier = 1 + (runCount - threshold) * 0.15;
  return baseMultiplier * additionalMultiplier;
}

/**
 * Get position for a player
 * TODO: Implement actual player lookup
 */
function getPlayerPosition(playerId: string): string {
  // Placeholder - should query player database
  return 'RB';
}

File: python/src/algorithms/position_run_detection.py
from typing import List
from src.models.types import DraftPick, PositionRun
from src.config import config


def detect_position_run(
    recent_picks: List[DraftPick], threshold: int = None
) -> PositionRun:
    """
    Detect if a position run is occurring
    A run is detected when 3+ consecutive picks are the same position
    """
    if threshold is None:
        threshold = config.POSITION_RUN_THRESHOLD

    if len(recent_picks) < threshold:
        return PositionRun(detected=False)

    last_n = recent_picks[-threshold:]
    positions = [get_player_position(p.player_id) for p in last_n]

    # Find most common position in recent picks
    position_counts = {}
    for pos in positions:
        position_counts[pos] = position_counts.get(pos, 0) + 1

    most_common = max(position_counts, key=position_counts.get)

    if position_counts[most_common] >= threshold:
        return PositionRun(
            detected=True, position=most_common, count=position_counts[most_common]
        )

    return PositionRun(detected=False)


def adjust_scarcity(position: str, run_count: int) -> float:
    """
    Adjust scarcity multiplier based on position run
    +15% per pick beyond threshold
    """
    base_multiplier = config.SCARCITY_MULTIPLIERS.get(position, 1.0)
    threshold = config.POSITION_RUN_THRESHOLD

    if run_count <= threshold:
        return base_multiplier

    additional_multiplier = 1 + (run_count - threshold) * 0.15
    return base_multiplier * additional_multiplier


def get_player_position(player_id: str) -> str:
    """
    Get position for a player
    TODO: Implement actual player lookup
    """
    # Placeholder - should query player database
    return "RB"
